### PR TITLE
Use Layer Shell and Qt6

### PR DESCRIPTION
**Pull request by stefonarch**
Mostly working, included all settings. Issues left are
* ~Tray menu left click opens fullscreen and transparent (due to overlay layer); at a second click it will reduce to normal size but is at top left of the screen.~ Disabled left click under wayland.
* ~It doesn't close loosing focus, needing Esc or action.~ Only left click.
*  Multiple notifications at bottom-left, bottom-right and bottom-center are going out of screen, not inverting as under x11.
* app_id and title are lost (not important as no window rules are needed)

screen_area_ven_16:49:06_

### CODE REVIEW

**Comment by tsujan** [comment]
Hopefully all issues will disappear with layer-shell-qt6, although the code might be very different.

**Comment by stefonarch** [comment]
Just found that right click on the icon in the tray works fine, it's the left click which opens fullscreen.

File: src/notificationarea.cpp [hunk 1]
```diff
@@ -168,4 +169,60 @@ void NotificationArea::setSettings(const QString &placement, int width, int spac
...
+                    layershell->setAnchors(anchors);
+                } else if (QL1S("bottom-center") == m_placement)
+                {
+                    LayerShellQt::Window::Anchors anchors = {LayerShellQt::Window::AnchorBottom};
+                    layershell->setAnchors(anchors);
+                } else if (QL1S("bottom-right") == m_placement)
+                {
+                    LayerShellQt::Window::Anchors anchors = {LayerShellQt::Window::AnchorBottom|                LayerShellQt::Window::AnchorRight};
+                    layershell->setAnchors(anchors);
+                }
```
Comment (tsujan) [inline src/notificationarea.cpp:222]:
```c++
               LayerShellQt::Window::Anchors anchors;
                if (m_placement == QL1S("top-left"))
                    anchors = {LayerShellQt::Window::AnchorTop|LayerShellQt::Window::AnchorLeft};
                else if (m_placement == QL1S("top-center"))
                    anchors = {LayerShellQt::Window::AnchorTop};
                else if (m_placement == QL1S("top-right"))
                    anchors = {LayerShellQt::Window::AnchorTop | LayerShellQt::Window::AnchorRight};
                // the other conditions come here......
                layershell->setAnchors(anchors);
```

File: src/notificationarea.cpp [hunk 2]
```diff
@@ -168,4 +169,60 @@ void NotificationArea::setSettings(const QString &placement, int width, int spac
...
+    if (underWayland)
+    {
+        winId();
+        if(QWindow* win = windowHandle())
+        {
+            if(LayerShellQt::Window* layershell = LayerShellQt::Window::get(win))
+            {
+                layershell->setLayer(LayerShellQt::Window::Layer::LayerOverlay);
+                layershell->setKeyboardInteractivity(LayerShellQt::Window::KeyboardInteractivityNone);
+                layershell->setMargins(margins);
```
[the remark this answers is not in the document]
Comment (stefonarch) [reply src/notificationarea.cpp:187]: Sorry I don't understand this.

File: src/notificationarea.cpp (hunk 1 above)
Comment (stefonarch) [reply src/notificationarea.cpp:222]: This gives meduring compiling several times:```

/tmp/lxqt-notificationd/src/notificationarea.cpp:191:51: warning: variable ‘anchors’ set but not used [-Wunused-but-set-variable]
  191 |                     LayerShellQt::Window::Anchors anchors = {LayerShellQt::Window::AnchorTop|LayerShellQt::Window::AnchorLeft};

```

and the last one, bottom right, is only working.

File: src/notificationarea.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsujan) [reply src/notificationarea.cpp:187]:
```c++
                layershell->setKeyboardInteractivity(LayerShellQt::Window::KeyboardInteractivityNone);
                QMargins margins(m_spacing, m_spacing, m_spacing, m_spacing);
                layershell->setMargins(margins);
```

File: src/notificationarea.cpp (hunk 1 above)
Comment (tsujan) [reply src/notificationarea.cpp:222]: Perhaps because you repeated `LayerShellQt::Window::Anchors`.

File: src/notification.cpp [hunk 3]
```diff
@@ -339,12 +340,12 @@ void Notification::mouseReleaseEvent(QMouseEvent * event)
             QWidget::mouseReleaseEvent(event);
             return;
         }
-        if (appName == appLabel->text() || windowTitle == appLabel->text())
-        {
-            KWindowSystem::raiseWindow(i);
-            closeButton_clicked();
-            return;
-        }
+ //       if (appName == appLabel->text() || windowTitle == appLabel->text())
```
Comment (stefonarch) [inline src/notification.cpp:343]: This I wasn't able to fix, `activateWindow` wasn't accepted.

**Comment by stefonarch** [comment]
Maybe this has to be rewritten (without layershell code it compiles)  it stops with:

```
[ 95%] Linking CXX executable lxqt-notificationd
/usr/bin/ld: CMakeFiles/lxqt-notificationd.dir/notificationarea.cpp.o: in function `NotificationArea::setSettings(QString const&, int, int, int, bool, QList<QString> const&)':
/home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:177:(.text+0xe1d): undefined reference to `LayerShellQt::Shell::useLayerShell()'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:183:(.text+0xe6c): undefined reference to `LayerShellQt::Window::get(QWindow*)'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:185:(.text+0xe8c): undefined reference to `LayerShellQt::Window::setLayer(LayerShellQt::Window::Layer)'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:186:(.text+0xe9d): undefined reference to `LayerShellQt::Window::setKeyboardInteractivity(LayerShellQt::Window::KeyboardInteractivity)'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:187:(.text+0xeb0): undefined reference to `LayerShellQt::Window::setMargins(QMargins const&)'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:192:(.text+0xf03): undefined reference to `LayerShellQt::Window::setAnchors(QFlags<LayerShellQt::Window::Anchor>)'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:197:(.text+0xf56): undefined reference to `LayerShellQt::Window::setAnchors(QFlags<LayerShellQt::Window::Anchor>)'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:201:(.text+0xfae): undefined reference to `LayerShellQt::Window::setAnchors(QFlags<LayerShellQt::Window::Anchor>)'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:205:(.text+0x1006): undefined reference to `LayerShellQt::Window::setAnchors(QFlags<LayerShellQt::Window::Anchor>)'
/usr/bin/ld: /home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:209:(.text+0x105e): undefined reference to `LayerShellQt::Window::setAnchors(QFlags<LayerShellQt::Window::Anchor>)'
/usr/bin/ld: CMakeFiles/lxqt-notificationd.dir/notificationarea.cpp.o:/home/stef/Neon/lxqt-notificationd/src/notificationarea.cpp:213: more undefined references to `LayerShellQt::Window::setAnchors(QFlags<LayerShellQt::Window::Anchor>)' follow
collect2: error: ld returned 1 exit status
```

**Comment by stefonarch** [comment]
It compiles, but now only the first notification is correct,  the following ones have weird positions and steal focus.

**Comment by tsujan** [comment]
OK, the problem was that we should *hide* the notification area instead of closing it under Wayland. The reason is that  `QWidget::close()` has been changed in Qt6.

**Comment by stefonarch** [comment]
All fine now except opening "Options" from systray: fullscreen and closing only with Esc.

**Comment by tsujan** [comment]
> opening "Options" from systray: fullscreen

That will be fixed by setting shell before showing the notification, not inside `NotificationArea::setSettings`.

File: src/notification.cpp (hunk 3 above)
Comment (tsujan) [reply src/notification.cpp:343]: Try to change it to

```c++
        if (appName == appLabel->text() || windowTitle == appLabel->text())
        {
            if (auto w = QWidget::find(i))
                w->raise(i);
            closeButton_clicked();
            return;
        }
```

File: src/notificationarea.h [hunk 4]
```diff
@@ -35,6 +35,11 @@
 /*! Top level widget. Scroll area is used to ensure access
  * of all \c Notification instances (scrollable with mouse).
  */
+
+/*namespace LayerShellQt {
+class Window
+}*/
+
```
Comment (tsujan) [inline src/notificationarea.h:42]: You can remove these lines.

File: src/notificationlayout.cpp [hunk 5]
```diff
@@ -52,7 +52,7 @@ NotificationLayout::NotificationLayout(QWidget *parent)
     setAttribute(Qt::WA_TranslucentBackground);
 
     m_layout = new QVBoxLayout(this);
-    m_layout->setMargin(0);
+    m_layout->setSpacing(0);
```
Comment (tsujan) [inline src/notificationlayout.cpp:55]: Please change it to

```c++
m_layout->setContentsMargins(0, 0, 0, 0);
```

**Comment by tsujan** [comment]
It seems that lxqt-notificationd is ready for both Qt6 and Wayland now. Thanks!

**Comment by stefonarch** [comment]
> It seems that lxqt-notificationd is ready for both Qt6 and Wayland now. Thanks!

With your precious help yes :)

**Comment by stefonarch** [comment]
To do: center config window.
https://github.com/lxqt/lxqt-config/issues/984

**Comment by tsujan** [comment]
> To do: center config window.

As I mentioned elsewhere, I don't agree.

**Review by tsujan — APPROVED** [review]

**Pull request merged**